index.js

```diff
@@ -30,21 +30,42 @@ var truncate = require('truncate');
  * @returns {Buffer}
  */
 function decode(uri) {
-  if (!/^data:/.test(uri))
+
+  //  dataurl    := "data:" [ mediatype ] [ ";base64" ] "," data
+  //  mediatype  := [ type "/" subtype ] *( ";" parameter )
+  //  data       := *urlchar
+  //  parameter  := attribute "=" value
+
+  var m = /^data:([^;,]+)?((?:;(?:[^;,]+))*?)(;base64)?,(.*)/.exec(uri);
+  if (!m) {
     throw new Error('Not a data URI: "' + truncate(uri, 20) + '"');
+  }
 
-  var commaIndex = uri.indexOf(',');
-  if (commaIndex < 0)
-    throw new Error('Invalid data URI "' + truncate(uri, 20) + '"');
+  var media  = '';
+  var b64    = m[3];
+  var body   = m[4];
+  var result = null;
 
-  var header = uri.slice(0, commaIndex);
-  var body = uri.slice(commaIndex+1);
+  // If <mediatype> is omitted, it defaults to text/plain;charset=US-ASCII.
+  // As a shorthand, "text/plain" can be omitted but the charset parameter
+  // supplied.
+  if (m[1]) {
+    media = m[1] + (m[2] || '');
+  } else {
+    if (m[2]) {
+      media = 'text/plain' + m[2];
+    } else {
+      media = 'text/plain;charset=US-ASCII';
+    }
+  }
 
-  if (/;base64$/.test(header)) {
-    return new Buffer(body, 'base64');
+  if (b64) {
+    result = new Buffer(body, 'base64');
   } else {
-    return new Buffer(decodeURIComponent(body), 'ascii');
+    result = new Buffer(decodeURIComponent(body), 'ascii');
   }
+  result.mediatype = media;
```

Comment:
since you have already parsed the charset, can you please expose it on the result too? For example as `result.charset`. It's useful for consumers that want to convert the result buffer into a string.

```
var result = dataUri.decode('data:text/plain;base64,aGVsbG8gd29ybGQ=');
var str = result.toString(result.charset);
```